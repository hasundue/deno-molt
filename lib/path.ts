import { dirname, fromFileUrl, join, resolve, toFileUrl } from "./std/path.ts";

/**
 * Convert a path to a string URL
 */
export function toUrl(path: string | URL): string {
  if (path instanceof URL) {
    return path.href;
  }
  // Exclude Windows paths like "C:\foo\bar"
  if (URL.canParse(path) && !path.match(/^[a-zA-Z]:/)) {
    return path;
  }
  // Assume the path is a relative path from the current working directory.
  return toFileUrl(resolve(path)).href;
}

/**
 * Convert a path to an absolute file path or a string URL.
 */
export function toPath(path: string | URL): string {
  if (path instanceof URL) {
    return path.protocol === "file:" ? fromFileUrl(path) : path.href;
  }
  if (URL.canParse(path)) {
    return path.startsWith("file:") ? fromFileUrl(path) : path;
  }
  return resolve(path);
}

/**
 * Recursively searches for file(s) with the specified name in parent directories
 * starting from the given starting directory.
 *
 * @param dir - The path to the directory to start searching from.
 * @param files - The name of the files to search for.
 * @returns The first file path found or undefined if no file was found.
 */
export async function findFileUp(dir: string | URL, ...files: string[]) {
  dir = toPath(dir);
  for (;;) {
    for await (const dirEntry of Deno.readDir(dir)) {
      if (files.includes(dirEntry.name)) {
        return join(dir, dirEntry.name);
      }
    }
    const newDir = dirname(dir);
    if (newDir === dir) {
      // reached the system root
      return undefined;
    }
    dir = newDir;
  }
}
