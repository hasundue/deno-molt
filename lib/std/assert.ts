export { assertArrayIncludes } from "https://deno.land/std@0.207.0/assert/assert_array_includes.ts";
export { assertEquals } from "https://deno.land/std@0.207.0/assert/assert_equals.ts";
export { assertExists } from "https://deno.land/std@0.207.0/assert/assert_exists.ts";
export { assertNotEquals } from "https://deno.land/std@0.207.0/assert/assert_not_equals.ts";
export { assertObjectMatch } from "https://deno.land/std@0.207.0/assert/assert_object_match.ts";
export { assertThrows } from "https://deno.land/std@0.207.0/assert/assert_throws.ts";
export { AssertionError } from "https://deno.land/std@0.207.0/assert/assertion_error.ts";
export { assertRejects } from "https://deno.land/std@0.207.0/assert/assert_rejects.ts";
