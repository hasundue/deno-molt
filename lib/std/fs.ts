export { existsSync } from "https://deno.land/std@0.208.0/fs/exists.ts";
export {
  detect as detectEOL,
  EOL,
  format as formatEOL,
} from "https://deno.land/std@0.208.0/fs/eol.ts";
