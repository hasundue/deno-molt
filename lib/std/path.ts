export { isAbsolute } from "https://deno.land/std@0.219.1/path/is_absolute.ts";
export { relative } from "https://deno.land/std@0.219.1/path/relative.ts";
export { resolve } from "https://deno.land/std@0.219.1/path/resolve.ts";
export { toFileUrl } from "https://deno.land/std@0.219.1/path/to_file_url.ts";
export { fromFileUrl } from "https://deno.land/std@0.219.1/path/from_file_url.ts";
export { extname } from "https://deno.land/std@0.219.1/path/extname.ts";
export { join } from "https://deno.land/std@0.219.1/path/join.ts";
export { dirname } from "https://deno.land/std@0.219.1/path/dirname.ts";
export { basename } from "https://deno.land/std@0.219.1/path/basename.ts";
