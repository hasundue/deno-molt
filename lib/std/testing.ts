export {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "https://deno.land/std@0.204.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.204.0/testing/snapshot.ts";
export {
  assertSpyCall,
  assertSpyCallArg,
  assertSpyCalls,
  type ConstructorSpy,
  type ExpectedSpyCall,
  type Spy,
  spy,
  type SpyCall,
  type Stub,
  stub,
} from "https://pax.deno.dev/hasundue/deno_std@feat-constructor-spy/testing/mock.ts";
