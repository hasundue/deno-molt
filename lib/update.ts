import { distinct } from "./std/collections.ts";
import {
  createGraph,
  type CreateGraphOptions,
  init as initDenoGraph,
  load as defaultLoad,
  type ModuleJson,
} from "./x/deno_graph.ts";
import { URI } from "./uri.ts";
import type { Maybe } from "./types.ts";
import { ImportMap, ImportMapJson } from "./import_map.ts";
import { Dependency, type DependencyProps, parseSemVer } from "./dependency.ts";

type DependencyJson = NonNullable<ModuleJson["dependencies"]>[number];

export type VersionProp = {
  from?: string;
  to: string;
};

/** Representation of a dependency update. */
export interface DependencyUpdate extends Omit<DependencyProps, "version"> {
  /** The fully resolved specifier of the dependency. */
  specifier: {
    from: URI<"http" | "https" | "npm">;
    to: URI<"http" | "https" | "npm">;
  };
  version: VersionProp;
  /** The code of the dependency. */
  code?: {
    /** The original specifier of the dependency appeared in the code. */
    specifier: string;
    span: NonNullable<DependencyJson["code"]>["span"];
  };
  type?: {
    specifier: string;
    span: NonNullable<DependencyJson["type"]>["span"];
  };
  /** The specifier of the module that imports the dependency. */
  referrer: URI<"file">;
  /** Information about the import map used to resolve the dependency. */
  map?: {
    /** The path to the import map used to resolve the dependency. */
    source: URI<"file">;
    from: string;
    /** The string in the dependency specifier being replaced by the import map.
     * Mapping on a file specifier should not happen. */
    to: URI<"http" | "https" | "npm">;
  };
}

export const DependencyUpdate = {
  collect,
  applyToModule,
  applyToImportMap,
};

class DenoGraph {
  static #initialized = false;

  static async ensureInit() {
    if (this.#initialized) {
      return;
    }
    await initDenoGraph();
    this.#initialized = true;
  }
}

const load: NonNullable<CreateGraphOptions["load"]> = async (
  specifier,
) => {
  const url = new URL(specifier); // should not throw
  switch (url.protocol) {
    case "node:":
    case "npm:":
      return {
        kind: "external",
        specifier,
      };
    case "http:":
    case "https:":
      return {
        kind: "external",
        specifier,
      };
    case "file:":
      return await defaultLoad(specifier);
    default:
      throw new Error(`Unsupported protocol: ${url.protocol}`);
  }
};

export async function collect(
  entrypoints: string | string[],
  options: {
    importMap?: string;
  } = {},
): Promise<DependencyUpdate[]> {
  // This could throw if the entrypoints are not valid URIs.
  const specifiers = [entrypoints].flat().map((path) => URI.from(path));

  // Ensure the deno_graph WASM module is initialized.
  await DenoGraph.ensureInit();

  const importMap = options.importMap
    ? await ImportMap.readFromJson(URI.from(options.importMap))
    : undefined;

  const graph = await createGraph(specifiers, {
    load,
    resolve: importMap ? importMap.resolveSimple : undefined,
  });

  const updates: DependencyUpdate[] = [];
  await Promise.all(
    graph.modules.flatMap((module) =>
      module.dependencies?.map(async (dependency) => {
        const update = await _create(
          dependency,
          URI.from(module.specifier),
          { importMap },
        );
        return update ? updates.push(update) : undefined;
      })
    ),
  );
  return updates;
}

export async function _create(
  dependency: DependencyJson,
  referrer: URI<"file">,
  options?: {
    importMap?: ImportMap;
    force?: boolean;
  },
): Promise<DependencyUpdate | undefined> {
  const specifier = dependency.code?.specifier ?? dependency.type?.specifier;
  if (!specifier) {
    throw new Error(
      `The dependency ${dependency.specifier} in ${
        URI.relative(referrer)
      } has no resolved specifier.`,
    );
  }
  const latest = await Dependency.resolveLatestURL(new URL(specifier));
  if (!latest) {
    return;
  }
  const props = Dependency.parseProps(latest);
  const mapped = options?.importMap?.resolve(
    dependency.specifier,
    referrer,
  );
  return {
    ...props,
    // We prefer to put the fully resolved specifier here.
    specifier: {
      from: URI.ensure("http", "https", "npm")(specifier),
      to: URI.ensure("http", "https", "npm")(latest.href),
    },
    code: dependency.code && {
      // We prefer to put the original specifier here.
      specifier: dependency.specifier,
      span: dependency.code.span,
    },
    type: dependency.type && {
      // We prefer to put the original specifier here.
      specifier: dependency.specifier,
      span: dependency.type.span,
    },
    version: {
      from: parseSemVer(specifier),
      to: props.version!, // Latest URL must have a semver
    },
    referrer,
    map: mapped
      ? {
        source: options!.importMap!.specifier,
        from: mapped.from!,
        to: URI.ensure("http", "https", "npm")(mapped.to!),
      }
      : undefined,
  };
}

function applyToModule(
  /** The dependency update to apply. */
  update: DependencyUpdate,
  /** Content of the module to update. */
  content: string,
): string {
  const span = update.code?.span ?? update.type?.span!;
  if (span.start.line !== span.end.line) {
    throw new Error(
      `The import specifier ${update.specifier} in ${update.referrer} is not a single line`,
    );
  }
  const line = span.start.line;
  const lines = content.split("\n");
  lines[line] = lines[line].replace(update.specifier.from, update.specifier.to);
  return lines.join("\n");
}

export function applyToImportMap(
  /** The dependency update to apply. */
  update: DependencyUpdate,
  /** Content of the import map to update. */
  content: string,
): string {
  const json = JSON.parse(content) as ImportMapJson;
  json.imports[update.map!.from] = update.map!.to.replace(
    update.specifier.from,
    update.specifier.to,
  );
  return JSON.stringify(json, null, 2);
}

export function createVersionProp(
  dependencies: DependencyUpdate[],
): Maybe<VersionProp> {
  const modules = distinct(dependencies.map((d) => d.name));
  if (modules.length > 1) {
    // Cannot provide a well-defined version prop
    return;
  }
  const tos = distinct(dependencies.map((d) => d.version.to));
  if (tos.length > 1) {
    throw new Error(
      "Multiple target versions are specified for a single module",
    );
  }
  const froms = distinct(dependencies.map((d) => d.version.from));
  return {
    from: froms.length === 1 ? froms[0] : undefined,
    to: tos[0],
  };
}
