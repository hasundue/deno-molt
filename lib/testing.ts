import { createAssertSnapshot, spy, Stub, stub } from "./std/testing.ts";
import { EOL, formatEOL } from "./std/fs.ts";
import { fromFileUrl } from "./std/path.ts";

export const assertSnapshot = createAssertSnapshot({
  dir: fromFileUrl(new URL("../test/snapshots/", import.meta.url)),
});

export const CommandStub = {
  create(CommandSpy = spy(Deno.Command)) {
    return class CommandStub extends CommandSpy {
      #output: Deno.CommandOutput = {
        code: 0,
        stdout: new Uint8Array(),
        stderr: new Uint8Array(),
        success: true,
        signal: null,
      };
      outputSync() {
        return this.#output;
      }
      output() {
        return Promise.resolve(this.#output);
      }
      spawn() {
        return new Deno.ChildProcess();
      }
      static clear() {
        this.calls = [];
      }
    };
  },
};
export type CommandStub = ReturnType<typeof CommandStub.create>;

export class FileSystemFake extends Map<string | URL, string> {}

export const ReadTextFileStub = {
  create(
    fs: FileSystemFake,
    options?: {
      readThrough?: boolean;
    },
  ): Stub {
    const original = Deno.readTextFile;
    return stub(
      Deno,
      "readTextFile",
      async (path) => {
        return fs.get(path.toString()) ??
          (options?.readThrough
            ? await original(path)
            : _throw(new Deno.errors.NotFound(`File not found: ${path}`)));
      },
    );
  },
};
export type ReadTextFileStub = ReturnType<typeof ReadTextFileStub.create>;

export const WriteTextFileStub = {
  create(
    fs: FileSystemFake,
  ) {
    return stub(
      Deno,
      "writeTextFile",
      (path, data) => {
        fs.set(path.toString(), formatEOL(data.toString(), EOL.LF));
        return Promise.resolve();
      },
    );
  },
};
export type WriteTextFileStub = ReturnType<typeof WriteTextFileStub.create>;

export const FetchStub = {
  create(
    createResponse: (
      request: string | URL | Request,
      init: RequestInit & { original: typeof fetch },
    ) => Response | Promise<Response>,
  ) {
    const original = globalThis.fetch;
    return stub(
      globalThis,
      "fetch",
      (request, init) =>
        Promise.resolve(createResponse(request, { ...init, original })),
    );
  },
};
export type FetchStub = ReturnType<typeof FetchStub.create>;

export const LatestSemVerStub = {
  create(latest: string): FetchStub {
    return FetchStub.create((request, init) => {
      request = (request instanceof Request)
        ? request
        : new Request(request, init);
      const url = new URL(request.url);
      switch (url.hostname) {
        case "registry.npmjs.org":
          return new Response(
            JSON.stringify({ "dist-tags": { latest } }),
            { status: 200 },
          );
        case "deno.land": {
          if (request.method !== "HEAD") {
            return init.original(request);
          }
          const { name, path } = parseDenoLandUrl(url);
          return {
            arrayBuffer: () => Promise.resolve(new ArrayBuffer(0)),
            redirected: true,
            status: 302,
            url: `https://${name}@${latest}${path}`,
          } as Response;
        }
        default:
          return init.original(request, init);
      }
    });
  },
};
export type LatestSemVerStub = ReturnType<typeof LatestSemVerStub.create>;

function parseDenoLandUrl(url: URL) {
  const std = url.pathname.startsWith("/std");
  const matched = std
    ? url.pathname.match(
      /^\/std(?:@(?<version>[^/]+))?(?<path>\/(.*)$)/,
    )
    : url.pathname.match(
      /^\/x\/(?<name>[^/]+)(?:@(?<version>[^/]+))?(?<path>\/(.*)$)/,
    );
  if (!matched) {
    throw new Error(`Unexpected URL: ${url}`);
  }
  const { name, version, path } = matched.groups!;
  return {
    name: std ? "deno.land/std" : `deno.land/x/${name}`,
    version,
    path,
  };
}

/**
 * Enables all test stubs.
 */
export function enableTestMode() {
  const fs = new FileSystemFake();
  ReadTextFileStub.create(fs, { readThrough: true });
  WriteTextFileStub.create(fs);
  LatestSemVerStub.create("123.456.789");
  Deno.Command = CommandStub.create();
}

/** Utility function to throw an error. */
function _throw(error: Error): never {
  throw error;
}
