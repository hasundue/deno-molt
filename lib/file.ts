import { DependencyUpdate } from "./update.ts";
import { URI } from "./uri.ts";

export interface FileUpdate {
  /** The specifier of the updated dependency (a remote module.) */
  specifier: URI<"file">;
  /** The updated content of the module. */
  content: string;
  /** The dependency updates in the module. */
  dependencies: DependencyUpdate[];
}

export const FileUpdate = {
  collect,
  write,
  writeAll,
};

function collect(
  dependencies: DependencyUpdate[],
): FileUpdate[] {
  /** A map of module specifiers to the module content updates. */
  const results = new Map<URI<"file">, FileUpdate>();
  for (const dependency of dependencies) {
    const referrer = dependency.map?.source ?? dependency.referrer;
    const current = results.get(referrer) ?? {
      specifier: referrer,
      content: Deno.readTextFileSync(new URL(referrer)),
      dependencies: [],
    } satisfies FileUpdate;
    const content = dependency.map
      ? DependencyUpdate.applyToImportMap(dependency, current.content)
      : DependencyUpdate.applyToModule(dependency, current.content);
    results.set(referrer, {
      specifier: current.specifier,
      content,
      dependencies: current.dependencies.concat(dependency),
    });
  }
  return Array.from(results.values());
}

export function writeAll(
  updates: FileUpdate[],
  options?: {
    onWrite?: (result: FileUpdate) => void;
  },
): void {
  updates.forEach((it) => {
    write(it);
    options?.onWrite?.(it);
  });
}

export function write(
  result: FileUpdate,
): void {
  Deno.writeTextFileSync(new URL(result.specifier), result.content);
}
