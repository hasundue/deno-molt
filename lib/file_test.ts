import { basename, dirname } from "./std/path.ts";
import { formatEOL, LF } from "./std/fs.ts";
import {
  FileSystemFake,
  ReadTextFileStub,
  WriteTextFileStub,
} from "./testing.ts";
import { assertInstanceOf } from "./std/assert.ts";
import { assertSnapshot } from "./testing.ts";
import * as DependencyUpdate from "./update.ts";
import { associateByFile, type FileUpdate, write } from "./file.ts";
import { LatestSemVerStub } from "./testing.ts";

const LATEST = "123.456.789";
LatestSemVerStub.create(LATEST);

function toName(path: string) {
  const base = basename(path);
  return base === "mod.ts" || base.endsWith(".json") || base.endsWith(".jsonc")
    ? basename(dirname(path)) + "/" + base
    : base;
}

async function assertFileUpdateSnapshot(
  t: Deno.TestContext,
  results: FileUpdate[],
) {
  await assertSnapshot(
    t,
    results.map((file) => ({
      dependencies: file.dependencies.map((it) => ({
        code: it.code,
        from: it.from,
        map: it.map
          ? {
            key: it.map.key,
            resolved: it.map.resolved,
          }
          : undefined,
        to: it.to,
      })),
      kind: file.kind,
    })),
  );
}

async function assertFileSystemSnapshot(
  t: Deno.TestContext,
  fs: FileSystemFake,
) {
  await assertSnapshot(
    t,
    Array.from(fs.entries()).map(([, content]) => formatEOL(content, LF)),
  );
}

const fs = new FileSystemFake();
ReadTextFileStub.create(fs, { readThrough: true });
WriteTextFileStub.create(fs);

async function test(path: string, name = toName(path)) {
  try {
    const updates = await DependencyUpdate.collect(
      new URL(path, import.meta.url),
      { cwd: new URL(dirname(path), import.meta.url) },
    );
    const results = associateByFile(updates);

    Deno.test("associateByFile - " + name, async (t) => {
      await assertFileUpdateSnapshot(t, results);
    });

    Deno.test("write - " + name, async (t) => {
      fs.clear();
      await write(results);
      await assertFileSystemSnapshot(t, fs);
    });
  } catch (error) {
    // import_map_reffered/deno.json just reffers to another import_map.json
    assertInstanceOf(error, SyntaxError);
  }
}

// Test the all cases in test/data
for await (
  const testCase of Deno.readDir(new URL("../test/data", import.meta.url))
) {
  if (testCase.isFile && testCase.name.endsWith(".ts")) {
    await test(`../test/data/${testCase.name}`);
  }
  if (testCase.isDirectory) {
    for await (
      const entry of Deno.readDir(
        new URL("../test/data/" + testCase.name, import.meta.url),
      )
    ) {
      if (
        entry.isFile && entry.name === "mod.ts" ||
        entry.name.endsWith(".json") || entry.name.endsWith(".jsonc")
      ) {
        await test(`../test/data/${testCase.name}/${entry.name}`);
      }
    }
  }
}
