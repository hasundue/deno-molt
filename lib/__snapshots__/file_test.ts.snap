export const snapshot = {};

snapshot[`writeAll > direct import 1`] = `
'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.59.1/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
'
`;

snapshot[`writeAll > direct import 2`] = `
'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
'
`;

snapshot[`writeAll > import map 1`] = `
'{
  "imports": {
    "std/": "https://deno.land/std@0.200.0/",
    "deno_graph": "https://deno.land/x/deno_graph@0.59.1/mod.ts",
    "node-emoji": "npm:node-emoji@2.1.0",
    "/": "./"
  }
}'
`;

snapshot[`writeAll > import map 2`] = `
'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
'
`;

snapshot[`writeAll > import map with no resolve 1`] = `
'import {} from "https://deno.land/std@0.204.0/version.ts";
'
`;

snapshot[`writeAll > unversioned specifiers 1`] = `
'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.59.1/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
'
`;

snapshot[`writeAll > unversioned specifiers 2`] = `
'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
'
`;
